bindings/c/src/api/transport.rs

```diff
@@ -0,0 +1,679 @@
+    use iota_streams::{
+        app::{
+            cty::{c_char, uint8_t, size_t, uint64_t},
+            cstr_core::{CStr, CString},
+            transport::tangle::client::{
+                SendOptions,
+                iota_client::{ClientBuilder, Api}
+            }
+        },
+        core::prelude::{ToString, String, Vec},
+    };
+    use core::time::Duration;
+
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    pub enum ClientApi {
+        GetHealth,
+        GetInfo,
+        GetPeers,
+        GetTips,
+        PostMessage,
+        PostMessageWithRemotePow,
+        GetOutput,
+        GetMilestone,
+        GetMessage,
+        GetBalance,
+    }
+
+    impl From<ClientApi> for Api {
+        fn from(api: ClientApi) -> Self {
+            match api {
+                ClientApi::GetHealth => Api::GetHealth,
+                ClientApi::GetInfo => Api::GetInfo,
+                ClientApi::GetPeers => Api::GetPeers,
+                ClientApi::GetTips => Api::GetTips,
+                ClientApi::PostMessage => Api::PostMessage,
+                ClientApi::PostMessageWithRemotePow => Api::PostMessageWithRemotePow,
+                ClientApi::GetOutput => Api::GetOutput,
+                ClientApi::GetMilestone => Api::GetMilestone,
+                ClientApi::GetMessage => Api::GetMessage,
+                ClientApi::GetBalance => Api::GetBalance,
+            }
+        }
+    }
+
+    #[derive(Default)]
+    pub struct ClientOptions<'a> {
+        node: Option<&'a str>,
+        primary_node: Option<(&'a str, Option<String>, Option<(&'a str, &'a str)>)>,
+        primary_pow_node: Option<(&'a str, Option<String>, Option<(&'a str, &'a str)>)>,
+        permanode: Option<(&'a str, Option<String>, Option<(&'a str, &'a str)>)>,
+        node_auth: Option<(&'a str, Option<String>, Option<(&'a str, &'a str)>)>,
+        nodes: Option<Vec<&'a str>>,
+        pool_urls: Option<Vec<String>>,
+        quorum: Option<bool>,
+        quorum_size: Option<usize>,
+        quorum_threshold: Option<usize>,
+        network: Option<&'a str>,
+        sync_interval: Option<u64>,
+        sync_disabled: bool,
+        offline_mode: bool,
+        request_timeout: Option<u64>,
+        api_timeout: Option<(ClientApi, u64)>,
+        local_pow: bool,
+        tips_interval: Option<u64>,
+    }
+
+    pub struct NodeList(Vec<String>);
+
+    #[no_mangle]
+    pub unsafe extern "C" fn new_node_list() -> *mut NodeList {
+        safe_into_mut_ptr(NodeList(Vec::new()))
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn add_node_to_list(list: *mut NodeList, node: *mut c_char) -> Err {
+        CString::from_raw(node).into_string().map_or(Err::BadArgument, |node| {
+            list.as_mut().map_or(Err::NullArgument, |list| {
+                list.0.push(node);
+                Err::Ok
+            })
+        })
+    }
+
+
+    /// Make a client using the predefined options
+    #[no_mangle]
+    pub unsafe extern "C" fn client_builder(
```

Comment:
You mentioned lifetime issues when following the builder pattern right? I haven't run this, but it does compile. Does this solves your lifetime issues, or they where somewhere else?:
```rust
    /// Make a default client builder to use for transport generation
    #[no_mangle]
    pub unsafe extern "C" fn client_builder_new() -> *mut ClientBuilder {
        safe_into_mut_ptr(ClientBuilder::new())
    }

    /// Adds an IOTA node by its URL.
    #[no_mangle]
    pub unsafe extern "C" fn client_builder_with_node(mut this: *mut ClientBuilder, node: *const c_char) -> Err {
        CStr::from_ptr(node).to_str().map_or(Err::BadArgument, |node| {
            this.as_mut().map_or(Err::NullArgument, |builder| {
                Box::from_raw(builder)
                    .with_node(node)
                    .map_or(Err::BadArgument, |new_builder| {
                        this = safe_into_mut_ptr(new_builder);
                        Err::Ok
                    })
            })
        })
    }

    /// Adds an IOTA node by its URL to be used as primary PoW node (for remote PoW), with optional jwt and or basic
    /// authentication
    #[no_mangle]
    pub unsafe extern "C" fn client_builder_with_primary_pow_node(
        mut this: *mut ClientBuilder,
        node: *const c_char,
        jwt: *const c_char,
        auth_name: *const c_char,
        auth_pswd: *const c_char,
    ) -> Err {
        if this.is_null() || node.is_null() {
            return Err::NullArgument;
        }
        use core::ops::Not;
        let args = CStr::from_ptr(node).to_str().and_then(|node| {
            jwt.is_null()
                .not()
                .then(|| CStr::from_ptr(jwt).to_str().map(String::from))
                .transpose()
                .and_then(|jwt| {
                    auth_name
                        .is_null()
                        .not()
                        .then(|| CStr::from_ptr(auth_name).to_str())
                        .transpose()
                        .and_then(|auth_name| {
                            auth_pswd
                                .is_null()
                                .not()
                                .then(|| CStr::from_ptr(auth_pswd).to_str())
                                .transpose()
                                .map(|auth_pswd| (node, jwt, auth_name, auth_pswd))
                        })
                })
        });

        let (node, jwt, auth_name, auth_pswd) = match args {
            Ok(args) => args,
            Err(_) => return Err::BadArgument,
        };

        let auth = auth_name.and_then(|auth_name| auth_pswd.map(|auth_pswd| (auth_name, auth_pswd)));

        Box::from_raw(this)
            .with_primary_pow_node(node, jwt, auth)
            .map_or(Err::OperationFailed, |new_builder| {
                this = safe_into_mut_ptr(new_builder);
                Err::Ok
            })
    }
   
    // ETC
```